maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java

```diff
@@ -387,6 +404,36 @@ private PluginVersionResult resolveFromProject( PluginVersionRequest request, Li
         return null;
     }
 
+    @SuppressWarnings( "unchecked" )
+    private Map<String, PluginVersionResult> getCache( SessionData data )
+    {
+        Map<String, PluginVersionResult> cache = ( Map<String, PluginVersionResult> ) data.get( CACHE_KEY );
+        while ( cache == null )
+        {
+            cache = new ConcurrentHashMap<>( 256 );
+            if ( data.set( CACHE_KEY, null, cache ) )
+            {
+                break;
+            }
+            cache = ( Map<String, PluginVersionResult> ) data.get( CACHE_KEY );
+        }
+        return cache;
+    }
+
+    private static String getKey( PluginVersionRequest request )
+    {
+        StringBuilder sb = new StringBuilder();
+        sb.append( request.getGroupId() );
+        sb.append( ':' );
+        sb.append( request.getArtifactId() );
+        for ( RemoteRepository repository : request.getRepositories() )
```

Comment:
Can you explain why you add the repo ids to the key?

Reply:
In theory, the result will depend on the input parameters of the requests.
I think that the way it's called, the repositories will always be the same for a given maven build, but this is not something known to this implementation afaik.

Reply:
I guess one problem is that the result also depends on the `Model` from the `PluginVersionRequest.getPom()` which is not included in the key.  I missed this point, but I haven't been able to find a real usage of it.

Reply:
It is the same for deps since repos can be set in a pom basis (let say gradle repo in a gradle submodule) but this practise triggers all resolutions to happen again since a repo can shadow another one so it looks correct to key the cache like that to me.

Reply:
I've slightly changed the behavior to hit the cache after a possible POM resolution.  The effect is basically the same since maven itself does not use POM resolution from the request.  I've kept the key to include the repositories, which sounds better given the actual request input params.

Reply:
Wouldn't it suffice to use the id from the POM/Model for the key? Do we care whether this plugin came from repo x or y?

Reply:
IIRC it happened in the past that the same GAV was binary different in 2 repo's (commons-lang?) If we need it, I don't think we should use the repository id, as it is maybe not unique enough (id=nexus might be used by several companies). The url from which the pom was downloaded should be unique.

Reply:
We don't have access to the URL. From the result only the repo.

Reply:
@rfscholte @michael-o I'll change the key to use a class that will contain groupId / artifactId / repositories and use a correct equals/hashCode implementation... (fwiw, and we do have access to the full repository definition as it's actually used for downloading the metadata...)